lib/Support/JSON.cpp

```diff
@@ -0,0 +1,134 @@
+//===- Json.cpp - Json Utilities --------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utilities for file system path handling, supplementing the ones from
+// llvm::sys::path.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/JSON.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace json = llvm::json;
+
+using namespace circt;
+using mlir::UnitAttr;
+
+// NOLINTBEGIN(misc-no-recursion)
+LogicalResult circt::convertAttributeToJSON(llvm::json::OStream &json,
+                                            Attribute attr) {
+  return TypeSwitch<Attribute, LogicalResult>(attr)
+      .Case<DictionaryAttr>([&](auto attr) {
+        json.objectBegin();
+        for (auto subAttr : attr) {
+          json.attributeBegin(subAttr.getName());
+          if (failed(convertAttributeToJSON(json, subAttr.getValue())))
+            return failure();
+          json.attributeEnd();
+        }
+        json.objectEnd();
+        return success();
+      })
+      .Case<ArrayAttr>([&](auto attr) {
+        json.arrayBegin();
+        for (auto subAttr : attr)
+          if (failed(convertAttributeToJSON(json, subAttr)))
+            return failure();
+        json.arrayEnd();
+        return success();
+      })
+      .Case<BoolAttr, StringAttr>([&](auto attr) {
+        json.value(attr.getValue());
+        return success();
+      })
+      .Case<IntegerAttr>([&](auto attr) -> LogicalResult {
+        // If the integer can be accurately represented by a double, print
+        // it as an integer. Otherwise, convert it to an exact decimal string.
+        const auto &apint = attr.getValue();
+        if (apint.getActiveBits() >= std::numeric_limits<int64_t>::digits)
```

Comment:
Should this be `std::numeric_limits<double>`, (+/- 1)?

Oh, maybe the comment is out-of-date.  Consider using `isIntN(64)` (or `isSignedIntN(64)`) (and regardless, fixing the comment if it doesn't match the code)?

Has this been tested at the boundaries, and for signed/unsigned values (see IntegerType's Signedness).  Looks like this only works for Signless?  Quick check by adding a test for this to `JSONTest.cpp` (thanks for adding a unit test!) suggests that, anyway.

Also, is this intended to support negative numbers? Looks like presently -1 doesn't work (Signed or Signless), AFAICT?

Also this may be a little too conservative, can use all 64 bits (failure here includes using 64)? But better to be conservative here than overlook something like a sign bit or something, so rather keep this as-is if there's doubt.

Reply:
Since the `JSONToAttribute` form assumes the Signedness, perhaps fail/warn if the type won't round-trip?  Not sure we have enough information with raw JSON otherwise.